ocaml/xapi/repository.ml

```diff
@@ -1024,3 +1024,129 @@ let get_pool_updates_in_json ~__context ~hosts =
+
+let assert_valid_guidances gs =
+  let eq l s = GuidanceStrSet.equal (set_of_list l) s in
+  match List.map (fun x -> guidance_of_string x) gs with
+  | [RebootHost]
+  | [EvacuateHost]
+  | [RestartToolstack]
+  | [RestartDeviceModel] -> ()
+  | l when eq l set1 -> ()
+  | l when eq l set2 -> ()
+  | l when eq l set3 -> ()
+  | l when eq l set4 -> ()
+  | l ->
+    let msg =
+      Printf.sprintf "Found wrong guidance(s) before applying updates: %s"
+        (String.concat ";" (List.map (fun x -> string_of_guidance x) l))
+    in
+    raise Api_errors.(Server_error (internal_error, [msg]))
+
+let apply_immediate_guidances ~__context ~host guidances =
+  try
+    let num_of_host = List.length (Db.Host.get_all ~__context) in
+    let open Client in
+    Helpers.call_api_functions ~__context (fun rpc session_id ->
+        match guidances with
+        | ["RebootHost"] ->
+          Client.Host.reboot ~rpc ~session_id ~host
+        | ["EvacuateHost"] -> ()
+        | ["RestartDeviceModel"] ->
+          Client.Host.restart_device_models ~rpc ~session_id ~self:host
+        | l when GuidanceStrSet.equal (GuidanceStrSet.of_list l) set1 ->
+          Client.Host.restart_agent ~rpc ~session_id ~host
+        | l when GuidanceStrSet.equal (GuidanceStrSet.of_list l) set2 ->
+          Client.Host.restart_device_models ~rpc ~session_id ~self:host;
+          Client.Host.restart_agent ~rpc ~session_id ~host
+        | l when GuidanceStrSet.equal (GuidanceStrSet.of_list l) set3 ->
+          (* Evacuating host restarts device models *)
+          if num_of_host = 1 then Client.Host.restart_device_models ~rpc ~session_id ~self:host;
+          ()
+        | l when GuidanceStrSet.equal (GuidanceStrSet.of_list l) set4 ->
+          (* Evacuating host restarts device models *)
+          if num_of_host = 1 then Client.Host.restart_device_models ~rpc ~session_id ~self:host;
+          Client.Host.restart_agent ~rpc ~session_id ~host
+        | l ->
+          let msg =
+            Printf.sprintf
+              "Found wrong guidance(s) after applying updates: %s" (String.concat ";" l)
+          in
+          raise Api_errors.(Server_error (internal_error, [msg])))
+  with e ->
+    let ref = Ref.string_of host in
+    error "applying immediate guidances on host ref='%s' failed: %s"
+      ref (ExnHelper.string_of_exn e);
+    raise Api_errors.(Server_error (apply_guidance_failed, [ref]))
+
+let with_host_in_updating ~__context ~host f =
+  Xapi_stdext_pervasives.Pervasiveext.finally
+    (fun () ->
+       Db.Host.add_to_other_config ~__context ~self:host
+           ~key:Xapi_globs.host_in_updating ~value:"true";
+       f ())
+    (fun () ->
+       Db.Host.remove_from_other_config ~__context ~self:host
+         ~key:Xapi_globs.host_in_updating)
+
+let apply_updates ~__context ~self ~host ~hash =
+  (* This function runs on master host *)
+  try
+    with_host_in_updating ~__context ~host (fun () ->
+        with_pool_repository (fun () ->
```

Comment:
Deep nesting could be avoided by:

```
with_host_in_updating ~__context ~host @@ fun () ->
with_pool_repository @@ fun () ->
```

The disadvantage is that the end of the functions would be less obvious.